sampo/scheduler/timeline/zone_timeline.py

```diff
@@ -0,0 +1,238 @@
+            # if state[current_start_idx].event_type == EventType.START \
+            #         and not self._match_status(state[current_start_idx].available_workers_count, required_status):
+            #     current_start_idx += 1
+            #     # if self._match_status(state[current_start_idx].available_workers_count, required_status)
+            #     current_start_time = state[current_start_idx].time
+            #     continue
+
+            # here we are outside the all intervals or inside the interval with right status
+            # if we are outside intervals, we can be in right or wrong status, so let's check it
+            # else we are inside the interval with right status so let
+
+            # we should count starts and ends on timeline prefix before the start_time
+            # if starts_count is equal to ends_count, start_time is out of all the zone usage intervals
+            # so we can change its status
+            # starts_count = len([v for v in state[:current_start_idx + 1] if v.event_type == EventType.START])
+            # ends_count = len([v for v in state[:current_start_idx + 1] if v.event_type == EventType.END])
+            # if starts_count == ends_count \
+            #         and not self._match_status(state[current_start_idx].available_workers_count, required_status):
+            #     # we are outside all intervals, so let's decide should
+            #     # we change zone status or go to the next checkpoint
+            #     old_status = state[current_start_idx].available_workers_count
+            #     # TODO Make this time calculation better: search the time slot for zone change before the start time
+            #     change_cost = self._config.time_costs[old_status, required_status]
+            #     prev_cpkt_idx = state.bisect_right(current_start_time - change_cost)
+            #     if prev_cpkt_idx == current_start_idx or prev_cpkt_idx >= len(state):
+            #         # we can change status before current_start_time
+            #         start_time_changed = current_start_time
+            #     else:
+            #         start_time_changed = state[prev_cpkt_idx].time + 1 + change_cost # current_start_time + change_cost
+            #
+            #     next_cpkt_idx = min(current_start_idx + 1, len(state) - 1)
+            #     next_cpkt_time = state[next_cpkt_idx].time
+            #     if (parent_time <= next_cpkt_time <= start_time_changed
+            #             and self._match_status(state[next_cpkt_idx].available_workers_count, required_status)):
+            #         # waiting until the next checkpoint is faster that change zone status
+            #         current_start_time = next_cpkt_time
+            #         current_start_idx += 1
+            #     else:
+            #         current_start_time = start_time_changed
+            #     # renewing the end index
+            #     end_idx = state.bisect_right(current_start_time + exec_time)
```

Comment:
rudiment comments

Reply:
this is the code that should be tested and added later